src/main/java/org/jenkinsci/plugins/workflow/support/actions/AnnotatedLogAction.java

```diff
@@ -0,0 +1,309 @@
+    public transient FlowNode node;
+
+    private AnnotatedLogAction(FlowNode node) {
+        if (!node.isActive()) {
+            throw new IllegalStateException("cannot start writing logs to a finished node " + node);
+        }
+        this.node = node;
+    }
+
+    @Override public void onLoad(FlowNode node) {
+        this.node = node;
+    }
+
+    private static byte[] prefix(FlowNode node) {
+        return prefix(node.getId());
+    }
+
+    private static byte[] prefix(String id) {
+        return (id + NODE_ID_SEP).getBytes(StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Scans {@link FlowExecutionOwner#getLog} for lines annotated as belonging to this node.
+     * <p>{@inheritDoc}
+     */
+    @Override public AnnotatedLargeText<? extends FlowNode> getLogText() {
+        ByteBuffer buf = new ByteBuffer();
+        try (InputStream whole = node.getExecution().getOwner().getLog(0); InputStream wholeBuffered = new BufferedInputStream(whole)) {
+            ByteArrayOutputStream2 baos = new ByteArrayOutputStream2();
+            byte[] prefix = prefix(node);
+            READ: while (true) {
+                int c = wholeBuffered.read();
+                if (c == -1) {
+                    break;
+                }
+                baos.write(c);
+                if (c == '\n') {
+                    byte[] linebuf = baos.getBuffer();
+                    int len = baos.size();
+                    if (len >= prefix.length) {
+                        boolean matches = true;
+                        for (int i = 0; i < prefix.length; i++) {
+                            if (linebuf[i] != prefix[i]) {
+                                matches = false;
+                                break;
+                            }
+                        }
+                        if (matches) {
+                            // This line in fact belongs to our node, so copy it out.
+                            buf.write(linebuf, prefix.length, len - prefix.length);
+                        }
+                    }
+                    baos.reset();
+                }
+            }
+        } catch (IOException x) {
+            LOGGER.log(Level.WARNING, null, x);
+        }
+        return new AnnotatedLargeText<>(buf, StandardCharsets.UTF_8, !node.isActive(), node);
+    }
+
+    // TODO probably need an API method in LogAction to obtain all log text from a block node and descendants (e.g., a parallel branch)
+
+    /**
+     * Used from <tt>console.jelly</tt> to write annotated log to the given output.
+     */
+    @Restricted(DoNotUse.class) // Jelly
+    public void writeLogTo(long offset, XMLOutput out) throws IOException {
+        getLogText().writeHtmlTo(offset, out.asWriter());
+    }
+
+    /**
+     * Creates a sink to print output from a step.
+     * Will use {@link LogActionImpl} if necessary.
+     * @param node a node which wishes to print output
+     * @param filter an optional log filter (must be {@link Serializable})
```

Comment:
As in https://github.com/jenkinsci/workflow-job-plugin/pull/27 and https://github.com/jenkinsci/workflow-api-plugin/pull/17 and https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/62#issuecomment-372345921 we should probably introduce into `remoting` something like

```java
public interface Remotable extends Serializable {
    default Channel getChannelForSerialization() throws NotSerializableException {/*…*/}
}
public interface SerializableOnlyOverRemoting extends Remotable {
    // empty
}
```